Add pyproject.toml for python packaging

JenySadadia opened the pull request:
Fixes https://github.com/kernelci/kernelci-core/issues/2283

With python 3.12 the support for setup.py is deprecated. Hence, update the packaging and distribution of 'kernelci' package by adding 'pyproject.toml' file.

yurinnick commented:
nit: If you simply moving files into `scripts` directory, `git mv` will do it better, preserving a history and keeping diff clean.

JenySadadia commented:
> nit: If you simply moving files into `scripts` directory, `git mv` will do it better, preserving a history and keeping diff clean.

Yes, that's true. But I didn't move them as it will affect current users and workflows as I am not sure if a specific run expects scripts to be in a specific location. @nuclearcat also flagged that maybe Jenkins workflow looks into `scripts/` directory for specific executables.

yurinnick commented:
I am not really opinionated about this, but it can be solved with `mv` + symlink in this case.

nuclearcat commented:
Small problem on staging:
```
Applying patch: patches/kernelci-core/staging.kernelci.org/0001-STAGING-add-build-configs-staging.yaml.patch
Applying patch: patches/kernelci-core/staging.kernelci.org/0002-STAGING-ensure-no-public-bisection-email-reports-are.patch
error: scripts/kci-bisect-push-results: wrong type
hint: Use 'git am --show-current-patch=diff' to see the failed patch
WARNING: Failed to apply patch
Aborting, all patches must apply.
```
adding skip label temporary to test other PR

JenySadadia commented:
> Small problem on staging:
> 
> ```
> Applying patch: patches/kernelci-core/staging.kernelci.org/0001-STAGING-add-build-configs-staging.yaml.patch
> Applying patch: patches/kernelci-core/staging.kernelci.org/0002-STAGING-ensure-no-public-bisection-email-reports-are.patch
> error: scripts/kci-bisect-push-results: wrong type
> hint: Use 'git am --show-current-patch=diff' to see the failed patch
> WARNING: Failed to apply patch
> Aborting, all patches must apply.
> ```
> 
> adding skip label temporary to test other PR

Thanks for reporting. I am taking a look.

JenySadadia commented:
Added `kernelci/script/update_sys_path.py` to enable imports from `kernelci` directory in `kci` scripts.

nuclearcat commented:
It still fails with same error, will add staging-skip as i'm debugging also problem with my patch on kubernetes

nuclearcat commented:
This is how it looks when it works:
```
Applying patch: patches/kernelci-core/staging.kernelci.org/0001-STAGING-add-build-configs-staging.yaml.patch
Applying patch: patches/kernelci-core/staging.kernelci.org/0002-STAGING-ensure-no-public-bisection-email-reports-are.patch
Applying patch: patches/kernelci-core/staging.kernelci.org/0003-STAGING-use-kernelci-staging-qemu-Docker-image.patch
Applying patch: patches/kernelci-core/staging.kernelci.org/0004-STAGING-use-staging.kernelci.org-branch-for-test-def.patch
Applying patch: patches/kernelci-core/staging.kernelci.org/0005-STAGING-use-staging-branch-for-buildroot-with-reduce.patch
Applying patch: patches/kernelci-core/staging.kernelci.org/0006-STAGING-use-staging-JWT-in-kubernetes-template.patch
```

JenySadadia commented:
> It still fails with same error, will add staging-skip as i'm debugging also problem with my patch on kubernetes

Yes, I haven't fixed it yet. I'm stuck with other issues in the PR.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  "toml==0.10.2",
+]
+authors = [
+  {name = "Guillaume Tucker" , email = "guillaume.tucker@collabora.com"},
```
Please change to `gtucker@gtucker.io`, this address is not in use anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Anders Roxell" , email = "anders.roxell@linaro.org"},
+  {name = "Ana Guerrero Lopez" , email = "ana.guerrero@collabora.com"},
+  {name = "Nícolas F. R. A. Prado" , email = "nfraprado@collabora.com"},
+  {name = "kernelci.org bot" , email = "bot@kernelci.org"},
```
Please drop this, the bot isn't an actual author.  Maybe listing the TSC group would be useful.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Kevin Hilman" , email = "khilman@baylibre.com"},
+  {name = "Matt Hart" , email = "matthew.hart@linaro.org"},
+  {name = "Jeny Sadadia" , email = "jeny.sadadia@collabora.com"},
+  {name = "Quentin Schulz" , email = "quentin.schulz@free-electrons.com"},
```
This one could probably be dropped, or updated with a bootlin address.  The free-electrons.com domain doesn't exist anymore (following a trademark dispute with Free ISP iirc around 2020).

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Michal Galka" , email = "michal.galka@collabora.com"},
+  {name = "Corentin LABBE" , email = "clabbe@baylibre.com"},
+  {name = "Anders Roxell" , email = "anders.roxell@linaro.org"},
+  {name = "Ana Guerrero Lopez" , email = "ana.guerrero@collabora.com"},
```
This one can probably be dropped too, at least the email isn't valid anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Tyler Baker" , email = "forcedinductionz@gmail.com"},
+  {name = "Laura Nao" , email = "laura.nao@collabora.com"},
+  {name = "Miguel Ojeda" , email = "ojeda@kernel.org"},
+  {name = "khilman" , email = "khilman@users.noreply.github.com"},
```
This looks like a dummy one that should be removed.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Alice Ferrazzi" , email = "alice.ferrazzi@miraclelinux.com"},
+  {name = "Adrian Ratiu" , email = "adrian.ratiu@collabora.com"},
+  {name = "Chris Paterson" , email = "chris.paterson2@renesas.com"},
+  {name = "Guillaume Tucker" , email = "gtucker@gtucker.io"},
```
Ah I see my current email address is here so the other @collabora.com entry can just be removed.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Chris Paterson" , email = "chris.paterson2@renesas.com"},
+  {name = "Guillaume Tucker" , email = "gtucker@gtucker.io"},
+  {name = "Lakshmipathi" , email = "lakshmipathi.ganapathi@collabora.com"},
+  {name = "Lakshmipathi Ganapathi" , email = "lakshmipathi.ganapathi@collabora.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Lakshmipathi Ganapathi" , email = "lakshmipathi.ganapathi@collabora.com"},
+  {name = "Khouloud Touil" , email = "ktouil@baylibre.com"},
+  {name = "Paweł Wieczorek" , email = "pawiecz@collabora.com"},
+  {name = "Enric Balletbo i Serra" , email = "enric.balletbo@collabora.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Aishwarya TCV" , email = "aishwarya.tcv@arm.com"},
+  {name = "Corentin Labbe" , email = "clabbe@baylibre.com"},
+  {name = "Ricardo Cañuelo" , email = "ricardo.canuelo@collabora.com"},
+  {name = "lakshmipathi.g" , email = "lakshmipathi.ganapathi@collabora.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Nicolas Belin" , email = "nbelin@baylibre.com"},
+  {name = "Dan Rue" , email = "dan.rue@linaro.org"},
+  {name = "Milo Casagrande" , email = "milo.casagrande@linaro.org"},
+  {name = "Tomeu Vizoso" , email = "tomeu.vizoso@collabora.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "lakshmipathi.g" , email = "lakshmipathi.ganapathi@collabora.com"},
+  {name = "Denis Efremov" , email = "efremov@linux.com"},
+  {name = "Jeny Sadadia" , email = "jeny.sadadia@gmail.com"},
+  {name = "Lakshmipathi Ganapathi" , email = "lakshmipathi.ganapathi@collabora.co.uk"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Florent Jacquet" , email = "florent.jacquet@free-electrons.com"},
+  {name = "Jan Luebbe" , email = "jlu@pengutronix.de"},
+  {name = "Michael Walle" , email = "michael@walle.cc"},
+  {name = "Milo Casagrande" , email = "milo@ubuntu.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Ed Smith" , email = "ed.smith@collabora.com"},
+  {name = "Heiko Thiery" , email = "heiko.thiery@gmail.com"},
+  {name = "Sjoerd Simons" , email = "sjoerd@collabora.com"},
+  {name = "Alex P" , email = "alexandra.pereira@collabora.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Sjoerd Simons" , email = "sjoerd@collabora.com"},
+  {name = "Alex P" , email = "alexandra.pereira@collabora.com"},
+  {name = "Detlev Casanova" , email = "detlev.casanova@collabora.com"},
+  {name = "Guillaume Tucker" , email = "guillaume@mangoz.org"},
```
Ha, that's a weird one.  I don't think I've used this address in about 10 years :p  So please drop it as well.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Ricardo Ribalda" , email = "ribalda@chromium.org"},
+  {name = "Chase Qi" , email = "chase.qi@linaro.org"},
+  {name = "Florian Bezdeka" , email = "florian.bezdeka@siemens.com"},
+  {name = "Kevin Hilman" , email = "khilman@kernel.org"},
+  {name = "Kevin Hilman" , email = "khilman@linaro.org"},
```
Probably each author should only appear once in this list.  It's not a list of email addresses but a list of people currently contributing to the code.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Florian Bezdeka" , email = "florian.bezdeka@siemens.com"},
+  {name = "Kevin Hilman" , email = "khilman@kernel.org"},
+  {name = "Kevin Hilman" , email = "khilman@linaro.org"},
+  {name = "Lakshmipathi.G" , email = "lakshmipathi.ganapathi@collabora.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Enric Balletbo i Serra" , email = "enric.balletbo@collabora.com"},
+  {name = "Nicolas Belin" , email = "nbelin@baylibre.com"},
+  {name = "Dan Rue" , email = "dan.rue@linaro.org"},
+  {name = "Milo Casagrande" , email = "milo.casagrande@linaro.org"},
```
Ditto - email doesn't exist anymore (foundries.io is his current address)

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Suram Chandra Sekhar" , email = "suram@oss.nxp.com"},
+  {name = "Tyler Baker" , email = "tyler@opensourcefoundries.com"},
+  {name = "William K Lin" , email = "lin.wen.kuang@gmail.com"},
+  {name = "alan bennett" , email = "akbennett@hasdell"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Tyler Baker" , email = "tyler@opensourcefoundries.com"},
+  {name = "William K Lin" , email = "lin.wen.kuang@gmail.com"},
+  {name = "alan bennett" , email = "akbennett@hasdell"},
+  {name = "isaiahduncan" , email = "32970507+isaiahduncan@users.noreply.github.com"},
```
Dummy entry.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "alan bennett" , email = "akbennett@hasdell"},
+  {name = "isaiahduncan" , email = "32970507+isaiahduncan@users.noreply.github.com"},
+  {name = "lollivier" , email = "lollivier@baylibre.com"},
+  {name = "Alexandra" , email = "alexandra@arch.localhost"},
```
Dummy entry.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "lollivier" , email = "lollivier@baylibre.com"},
+  {name = "Alexandra" , email = "alexandra@arch.localhost"},
+  {name = "Alexandre Belloni" , email = "alexandre.belloni@free-electrons.com"},
+  {name = "Alice Ferrazzi" , email = "alice.ferrazzi@cybertrust.co.jp"},
```
Replaced with Arisu.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Jakob Unterwurzacher" , email = "jakob.unterwurzacher@theobroma-systems.com"},
+  {name = "Jan-Simon Möller" , email = "dl9pf@gmx.de"},
+  {name = "Kees Cook" , email = "keescook@chromium.org"},
+  {name = "Lakshmipathi.G" , email = "Lakshmipathi.Ganapathi@Collabora.com"},
```
Ditto - email doesn't exist anymore.

It's impressive how many email addresses Laks has used :)

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Lakshmipathi.G" , email = "Lakshmipathi.Ganapathi@Collabora.com"},
+  {name = "Matt Hart" , email = "matthew.hart@arm.com"},
+  {name = "Matthias Maennich" , email = "maennich@google.com"},
+  {name = "Maxime Ripard" , email = "maxime.ripard@free-electrons.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Matt Hart" , email = "matthew.hart@arm.com"},
+  {name = "Matthias Maennich" , email = "maennich@google.com"},
+  {name = "Maxime Ripard" , email = "maxime.ripard@free-electrons.com"},
+  {name = "Mylène Josserand" , email = "mylene.josserand@collabora.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Maxime Ripard" , email = "maxime.ripard@free-electrons.com"},
+  {name = "Mylène Josserand" , email = "mylene.josserand@collabora.com"},
+  {name = "Neil Williams" , email = "codehelp@debian.org"},
+  {name = "Nick Desaulniers" , email = "nickdesaulniers@users.noreply.github.com"},
```
Dummy entry.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Sasha Levin" , email = "sashal@kernel.org"},
+  {name = "Shreeya Patel" , email = "shreeya.patel@collabora.com"},
+  {name = "Sjoerd Simons" , email = "sjoerd.simons@collabora.co.uk"},
+  {name = "Thomas Petazzoni" , email = "thomas.petazzoni@free-electrons.com"},
```
Ditto - email doesn't exist anymore.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+  {name = "Wang Mingyu" , email = "wangmy@cn.fujitsu.com"},
+  {name = "nkrishna" , email = "nkrishna@otc-prt.org"},
+  {name = "roxell" , email = "anders.roxell@gmail.com"},
+  {name = "sjoerd-ccu" , email = "sjoerd.simons@collabora.co.uk"}
```
There's already a Sjoerd Simmons entry so this one should be dropped.

gctucker commented on pyproject.toml:
```diff
@@ -0,0 +1,195 @@
+maintainers = [
+  {name = "Denys Fedoryshchenko", email = "denys.f@collabora.com"},
+  {name = "Jeny Sadadia", email = "jeny.sadadia@collabora.com"}
+]
```
Arisu is listed as co-maintainer too: https://kernelci.org/docs/org/maintainers/#core-tools

gctucker reviewed (changes requested):
A number of entries in the `authors` list aren't valid.  I think this is merely for users who install the package to know who is currently contributing to the code rather than a legally-binding list of contributors, which is covered by the headers in each file in the source code.  So probably it would be better to look at the top contributors or people who have contributed in the past year to manually curate this list into a more representative one.

JenySadadia commented:
> A number of entries in the `authors` list aren't valid. I think this is merely for users who install the package to know who is currently contributing to the code rather than a legally-binding list of contributors, which is covered by the headers in each file in the source code. So probably it would be better to look at the top contributors or people who have contributed in the past year to manually curate this list into a more representative one.

I have used `git shortlog --summary --email` to get all the developers and added them to this list.
Checking every file manually for the commit header would take a lot of time.

JenySadadia commented:
> > > > A number of entries in the `authors` list aren't valid. I think this is merely for users who install the package to know who is currently contributing to the code rather than a legally-binding list of contributors, which is covered by the headers in each file in the source code. So probably it would be better to look at the top contributors or people who have contributed in the past year to manually curate this list into a more representative one.
> > > 
> > > 
> > > I have used `git shortlog --summary --email` to get all the developers and added them to this list. Checking every file manually for the commit header would take a lot of time.
> > 
> > 
> > I don't think that's a valid approach, having this inaccurate list of contributors isn't helpful. I don't think it should take more than 15min to prepare by hand, in fact if you apply the changes from the PR review that would already be mostly done. I would suggest to not have a list of contributors at all if it's just a dump of the git history which isn't representative and contains lots of stale entries.
> 
> After taking a look at various popular Python projects, I've noticed that most of them don't specify a list of authors in `pyproject.toml`. Some have an `AUTHORS` file but that's really not required. So I think having a generic entry with the mailing list would be useful, and ensuring the list of maintainers and co-maintainers is accurate (so please add @aliceinwire to the list there as per my review comment).
> 
> It's also useful to take a look at how the legacy PyPI entry looks since this is what users will be seeing first with the v2.0 upload: https://pypi.org/project/kernelci/

Thanks for the reviews.
I have updated the list of maintainers and added Arisu to it.
About the authors, do you mean I should drop all the authors and just have `kernelci@groups.io` in the list?

gctucker commented:
Yes I think the list of authors should be dropped, and the new mailing list `kernelci@lists.linux.dev` should be mentioned somewhere in the meta-data but not necessarily as "authors".  One option is to just have this I guess:
```toml
authors = [
  {name = "KernelCI", email = "kernelci@lists.linux.dev"},
]
```
as I don't see any dedicated fields for a mailing list.  Django has something similar, except their have their own dedicated foundation:
https://pypi.org/project/Django/

JenySadadia commented:
> Yes I think the list of authors should be dropped, and the new mailing list `kernelci@lists.linux.dev` should be mentioned somewhere in the meta-data but not necessarily as "authors". One option is to just have this I guess:
> 
> ```toml
> authors = [
>   {name = "KernelCI", email = "kernelci@lists.linux.dev"},
> ]
> ```
> 
> as I don't see any dedicated fields for a mailing list. Django has something similar, except their have their own dedicated foundation: https://pypi.org/project/Django/

Okay. Thanks for the comment. Updated the authors' list.

JenySadadia commented on pyproject.toml:
```diff
@@ -0,0 +1,83 @@
+
+[project.scripts]
+kci = "kernelci.scripts.kci:main"
+kci_bisect = "kernelci.scripts.kci_bisect:main"
+kci_build = "kernelci.scripts.kci_build:main"
+kci_data = "kernelci.scripts.kci_data:main"
+kci_rootfs = "kernelci.scripts.kci_rootfs:main"
+kci_test = "kernelci.scripts.kci_test:main"
+kci_bisect_lava_v2_callback = "kernelci.scripts.kci_bisect_lava_v2_callback:main"
+kci_bisect_push_results = "kernelci.scripts.kci_bisect_push_results:main"
```
That's exactly why we have symlinks in the current location of the scripts e.g. `scripts/kci-bisect-push-results` 
It wasn't possible to use script names with `-` in the `project.scripts` sections. That's why I needed to rename the scripts.

gctucker replied:
Ah I see, well I didn't notice the symlinks.  Still I wonder if these scripts should actually be made part of the package, but I guess there isn't a clear deprecation timeline yet.

Alright so as long as Jenkins jobs aren't broken then it should be fine.  Let's just hope that moving to symlinks doesn't have unwanted side effects.

JenySadadia replied:
Tried to run Jenkins job https://bot.staging.kernelci.org/job/test-bisect/3/.
It wasn't able to kind `kci-bisect-lava-v2-callback` script. 
```
+ PYTHON_EGG_CACHE=/data/workspace/bot.staging.kernelci.org/kernel-builder/workspace/test-bisect/python-egg-cache kci-bisect-lava-v2-callback --token=**** --test-case=baseline.dmesg.emerg callback.json
/data/workspace/bot.staging.kernelci.org/kernel-builder/workspace/test-bisect__tmp/durable-171833ff/script.sh: 2: kci-bisect-lava-v2-callback: not found
```
Updated the PR to fix the issue.

JenySadadia replied:
Tested OK: https://bot.staging.kernelci.org/job/test-bisect/6/

nuclearcat commented:
Thanks Jeny, tested on staging, and thanks for @gctucker for hint to possible point of failure and review.
